File: scala/day17_part1_2024.scala

import scala.io.Source
import scala.util.Using

object Solution {
  def main(args: Array[String]): Unit = {
    Using(Source.fromFile("input.txt")) { source =>
      val lines = source.getLines().toList
      var A = 0
      var B = 0
      var C = 0
      var program = List[Int]()

      lines.foreach { line =>
        val trimmedLine = line.trim
        if (trimmedLine.startsWith("Register A:")) {
          A = trimmedLine.split(":").last.trim.toInt
        } else if (trimmedLine.startsWith("Register B:")) {
          B = trimmedLine.split(":").last.trim.toInt
        } else if (trimmedLine.startsWith("Register C:")) {
          C = trimmedLine.split(":").last.trim.toInt
        } else if (trimmedLine.startsWith("Program:")) {
          program = trimmedLine.split(":").last.trim.split(",").map(_.trim.toInt).toList
        }
      }

      def getComboVal(op: Int): Int = op match {
        case op if op <= 3 => op
        case 4 => A
        case 5 => B
        case 6 => C
        case _ => throw new IllegalArgumentException("invalid combo operand")
      }

      var outputVals = List[String]()
      var ip = 0
      while (ip < program.length) {
        val opcode = program(ip)
        if (ip + 1 >= program.length) {
          ip = program.length
        } else {
          val operand = program(ip + 1)
          opcode match {
            case 0 =>
              val den = getComboVal(operand)
              A = if (den == 0) 0 else A / (1 << den)
              ip += 2
            case 1 =>
              B = B ^ operand
              ip += 2
            case 2 =>
              B = getComboVal(operand) % 8
              ip += 2
            case 3 =>
              if (A != 0) ip = operand else ip += 2
            case 4 =>
              B = B ^ C
              ip += 2
            case 5 =>
              outputVals = outputVals :+ (getComboVal(operand) % 8).toString
              ip += 2
            case 6 =>
              val den = getComboVal(operand)
              B = A / (1 << den)
              ip += 2
            case 7 =>
              val den = getComboVal(operand)
              C = A / (1 << den)
              ip += 2
            case _ =>
              ip = program.length
          }
        }
      }
      println(outputVals.mkString(","))
    }
  }
}

File: kotlin/day17_part1_2024.kt

import java.io.File

fun main() {
    val lines = File("input.txt").readLines()
    var A = 0
    var B = 0
    var C = 0
    var program = emptyList<Int>()

    for (line in lines) {
        val trimmedLine = line.trim()
        if (trimmedLine.isEmpty()) continue
        when {
            trimmedLine.startsWith("Register A:") -> A = trimmedLine.substringAfter(":").trim().toInt()
            trimmedLine.startsWith("Register B:") -> B = trimmedLine.substringAfter(":").trim().toInt()
            trimmedLine.startsWith("Register C:") -> C = trimmedLine.substringAfter(":").trim().toInt()
            trimmedLine.startsWith("Program:") -> program = trimmedLine.substringAfter(":").trim().split(",").map { it.trim().toInt() }
        }
    }

    fun getComboVal(op: Int): Int = when (op) {
        in 0..3 -> op
        4 -> A
        5 -> B
        6 -> C
        else -> throw IllegalArgumentException("invalid combo operand")
    }

    val outputVals = mutableListOf<String>()
    var ip = 0
    while (ip < program.size) {
        val opcode = program[ip]
        if (ip + 1 >= program.size) break
        val operand = program[ip + 1]

        when (opcode) {
            0 -> {
                val den = getComboVal(operand)
                A = if (den == 0) 0 else A shr den
                ip += 2
            }
            1 -> {
                B = B xor operand
                ip += 2
            }
            2 -> {
                B = getComboVal(operand) % 8
                ip += 2
            }
            3 -> {
                ip = if (A != 0) operand else ip + 2
            }
            4 -> {
                B = B xor C
                ip += 2
            }
            5 -> {
                outputVals.add((getComboVal(operand) % 8).toString())
                ip += 2
            }
            6 -> {
                val den = getComboVal(operand)
                B = A shr den
                ip += 2
            }
            7 -> {
                val den = getComboVal(operand)
                C = A shr den
                ip += 2
            }
            else -> break
        }
    }
    println(outputVals.joinToString(","))
}
